index.bs

```diff
@@ -68,12 +71,145 @@ The appropriate times to <a spec=HTML>fetch and process the linked resource</a>
+
+* Rename [=the script's script=] to <dfn>the script's result</dfn>, which can be either a <a spec="html">script</a> or a [=speculation rule set=].
+
+The following algorithms are updated accordingly:
+
+* [=Prepare a script=]: see [[#speculation-rules-prepare-a-script-patch]].
+* <a spec=html>Execute a script block</a>: Add the following case to the switch on <a spec=html>the script's type</a>:
+  <dl>
+    <dt>"`speculationrules`"</dt>
+    <dd>
+      1. [=Assert=]: Never reached.
+    </dd>
+  </dl>
+
+<p class="issue">We should consider whether we also want to make this execute even if scripting is disabled.</p>
+
+<p class="issue">We should also incorporate the case where a {{HTMLScriptElement/src}} attribute is set.</p>
+
+<p class="issue">We could fire {{HTMLElement/error}} and {{HTMLElement/load}} events if we wanted to.</p>
+
+<h3 id="speculation-rules-prepare-a-script-patch">Prepare a script</h3>
+
+Inside the [=prepare a script=] algorithm we make the following changes:
+
+* Insert the following step as the second-last sub-step under "Determine the script's type as follows:":
+  * If the script block's type string is an [=ASCII case-insensitive=] match for the string "`speculationrules`", <a spec=html>the script's type</a> is "`speculationrules`".
+
+* Insert the following case in the switch on <a spec=html>the script's type</a> within the step which begins "If the element does not have a {{HTMLScriptElement/src}} content attribute..."
+  <dl>
+    <dt>"`speculationrules`"</dt>
+    <dd>
+      1. Let |result| be the result of [=parse speculation rules|parsing speculation rules=] given source text and base URL.
+
+      1. Set [=the script's result=] to |result|.
+
+      1. <a spec=html>The script is ready</a>.
+    </dd>
+  </dl>
+
+* Insert the following case to the switch in the subsequent step beginning "Then, follow the first of the following options...." after the cases which apply only to "`classic`" and "`module`" scripts:
+  <dl>
+    <dt>If <a spec=html>the script's type</a> is "`speculationrules`"</dt>
+    <dd>
+      1. When <a spec=html>the script is ready</a>, run the following steps:
+
+        1. If [=the script's result=] is not null, [=list/append=] it to the element's [=Node/node document=]'s [=document/list of speculation rule sets=].
+    </dd>
+  </dl>
+
+
+<h3 id="speculation-rules-parsing">Parsing</h3>
+
+<p class="note">
+  The general principle here is to allow the existence of directives which are not understood, but not to accept into the rule set a rule which the user agent does not fully understand.
+  This reduces the risk of unintended activity by user agents which are unaware of most recently added directives which might limit the scope of a rule.
+
+<div algorithm="parse speculation rules">
+  To <dfn export>parse speculation rules</dfn> given a [=string=] |input| and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule set=] or null.
```

Comment:
These exports aren't necessary, however I notice that they are also sprinkled throughout the import maps spec, so I'll defer to @domenic as for whether or not we should have these.

Reply:
Quite honestly I've been doing it for consistency with there, but I can certainly strip them. I have no idea whether we should be adding them by some criteria or only when/if some other spec wants to link o them.

Reply:
I've mostly added them in cases where otherwise Bikeshed would yell at me because nothing else in the spec references them. (Because, eventually in a non-monkeypatch spec, something _would_ reference them.) Otherwise I would not export them.

Reply:
I removed them.